tidb-binlog: add configuration file (#2064)
pingcap/docs: reference/tidb-binlog/configs.md

---
title: TiDB Binlog Configuration File
category: reference
---

# TiDB Binlog Configuration File

This document introduces the configuration items of TiDB Binlog.

## Pump

This section introduces the configuration items of Pump. For the example of a complete Pump configuration file, see [Pump Configuration](https://github.com/pingcap/tidb-binlog/blob/master/cmd/pump/pump.toml).

### addr

* Specifies the listening address of HTTP API in the format of `host: port`.

> * Default value: `"127.0.0.1:8250"`

Author: It is not valid. Removed in this PR, and I'll create another PR to modify the Chinese document.

* Specifies the log level.
* Default value: `"info"`

### node-id

* Specifies the Pump node ID. With this ID, this Pump process can be identified in the cluster.
* Default value: `hostname:port number`. For example, `node-1:8250`.

### security

This section introduces configuration items related to security.

#### ssl-ca

* Specifies the file path of the trusted SSL certificate list or CA list. For example, `/path/to/ca.pem`.
* Default value: `""`

#### ssl-cert

* Specifies the path of the X509 certificate file encoded in the Privacy Enhanced Mail (PEM) format. For example, `/path/to/pump.pem`.
* Default value: `""`

#### ssl-key

* Specifies the path of the X509 key file encoded in the PEM format. For example, `/path/to/pump-key.pem`.
* Default value: `""`

### storage

This section introduces configuration items related to storage.

#### sync-log

* Specifies whether to use `fsync` after each **batch** write to binlog to ensure data safety.
* Default value: `true`

#### kv_chan_cap

* Specifies the number of write requests that the buffer can store before Pump receives these requests.
* Default value: `1048576` (that is, 2 to the power of 20)

#### slow_write_threshold

> * The threshold (in seconds) If it takes longer to write a single binlog file than this specified threshold, the write is considered slow write and `"take a long time to write binlog"` is output in the log.

Reviewer: ```suggestion
* The threshold (in seconds) if it takes longer to write a single binlog file than this specified threshold, the write is considered slow write and `"take a long time to write binlog"` is output in the log.
```

> * The threshold (in seconds) if it takes longer to write a single binlog file than this specified threshold, the write is considered slow write and `"take a long time to write binlog"` is output in the log.

Reviewer 2: ```suggestion
* The threshold (in seconds). If it takes longer to write a single binlog file than this specified threshold, the write is considered slow write and `"take a long time to write binlog"` is output in the log.
```

* Default value: `1`

#### stop-write-at-available-space

* Binlog write requests is no longer accepted when the available storage space is below this specified value. You can use the format such as `900 MB`, `5 GB`, and `12 GiB` to specify the storage space. If there is more than one Pump node in the cluster, when a Pump node refuses a write request because of the insufficient space, TiDB will automatically write binlogs to other Pump nodes.
* Default value: `10 GiB`

#### kv

Currently the storage of Pump is implemented based on [GoLevelDB](https://github.com/syndtr/goleveldb). Under `storage` there is also a `kv` subgroup that is used to adjust the GoLevel configuration. The supported configuration items are shown as below:

* block-cache-capacity
* block-restart-interval
* block-size
* compaction-L0-trigger
* compaction-table-size
* compaction-total-size
* compaction-total-size-multiplier
* write-buffer
* write-L0-pause-trigger
* write-L0-slowdown-trigger

> For the detailed description of above items, see [GoLevelDB Document](https://godoc.org/github.com/syndtr/goleveldb/leveldb/opt#Options).

Reviewer 2: ```suggestion
For the detailed description of the above items, see [GoLevelDB Document](https://godoc.org/github.com/syndtr/goleveldb/leveldb/opt#Options).
```

## Drainer

This section introduces the configuration items of Drainer. For the example of a complete Drainer configuration file, see [Drainer Configuration](https://github.com/pingcap/tidb-binlog/blob/master/cmd/drainer/drainer.toml)

### addr

> * Specifies the listening address of HTTP API in the format of `host: port`.

Reviewer 2: ```suggestion
* Specifies the listening address of HTTP API in the format of `host:port`.
```

* Default value: `"127.0.0.1:8249"`

### advertise-addr

> * Specifies the externally accessible HTTP API address. This address is registered in PD in the format of `host: port`.

Reviewer 2: ```suggestion
* Specifies the externally accessible HTTP API address. This address is registered in PD in the format of `host:port`.
```

* Default value: `"127.0.0.1:8249"`

### log-file

* Specifies the path where log files are stored. If the parameter is set to an empty value, log files are not stored.
* Default value: `""`

### log-level

* Specifies the log level.
* Default value: `"info"`

### node-id

* Specifies the Drainer node ID. With this ID, this Drainer process can be identified in the cluster.
* Default value: `hostname:port number`. For example, `node-1:8249`.

### data-dir

* Specifies the directory used to store files that need to be saved during Drainer operation.
* Default value: `"data.drainer"`

### detect-interval

> * Specifies the interval (in seconds) at which PD updates the Pump information

Reviewer 2: ```suggestion
* Specifies the interval (in seconds) at which PD updates the Pump information.
```

* Default value: `5`

### pd-urls

* The comma-separated list of PD URLs. If multiple addresses are specified, the PD client will automatically attempt to connect to another address if an error occurs when connecting to one address.
* Default value: `"http://127.0.0.1:2379"`

### initial-commit-ts

> * Specifies from which commit timestamp the replication task starts. This configuration is only applicable to the Drainer node that started replication for the first time. If a checkpoint already exists downstream, the replication will be performed according to the time recorded in the checkpoint.

Reviewer 2: ```suggestion
* Specifies from which commit timestamp the replication task starts. This configuration is only applicable to the Drainer node that starts replication for the first time. If a checkpoint already exists downstream, the replication will be performed according to the time recorded in the checkpoint.
```

> * Default value: `-1`。Drainer 会从 PD 得到一个最新的 timestamp 作为初始时间。Drainer will get a new timestamp from PD as the starting time.

Reviewer: have Chinese

### synced-check-time

> * You can access the `/ status` path through the HTTP API to query the status of Drainer replication. `synced-check-time` specifies how many minutes from the last successful replication is considered as `synced`, that is, the replication is complete.

Reviewer 2: ```suggestion
* You can access the `/status` path through the HTTP API to query the status of Drainer replication. `synced-check-time` specifies how many minutes from the last successful replication is considered as `synced`, that is, the replication is complete.
```

* Default value: `"INFORMATION_SCHEMA,PERFORMANCE_SCHEMA,mysql"`

#### ignore-table

Ignores the specified table changes during replication. You can specify multiple tables to be ignored in the `toml` file. For example:

{{< copyable "" >}}

```toml
[[syncer.ignore-table]]
db-name = "test"
tbl-name = "log"

[[syncer.ignore-table]]
db-name = "test"
tbl-name = "audit"
```

If all changes in a binlog file are filtered, the whole binlog file is ignored.

Default value: `[]`

#### replicate-do-db

* Specifies the database to be replicated. For example, `[db1, db2]`.
* Default value: `[]`

#### replicate-do-table

Specifies the table to be replicated. For example:

{{< copyable "" >}}

```toml
[[syncer.replicate-do-table]]
db-name ="test"
tbl-name = "log"

[[syncer.replicate-do-table]]
db-name ="test"
tbl-name = "~^a.*"
```

Default value: `[]`

#### txn-batch

* When the downstream is the `mysql` or `tidb` type, DML operations are executed in different batches. This parameter specifies how many DML operations can be included in each transaction.
* Default value: `20`

#### worker-count

> * When the downstream is the `mysql` or `tidb` type, DML operations are executed in different batches. This parameter specifies the concurrency numbers of DML operations.

Reviewer 2: ```suggestion
* When the downstream is the `mysql` or `tidb` type, DML operations are executed concurrently. This parameter specifies the concurrency numbers of DML operations.
```

* Default value: `16`

#### disable-dispatch

> * Disables the concurrency and forcibly set `work-count` to `1`.

Reviewer 2: ```suggestion
* Disables the concurrency and forcibly set `worker-count` to `1`.
```

* Default value: `false`

#### safe-mode

> If Safe mode is enabled, Drainer modifies the replication updates in the following way:

Reviewer 2: ```suggestion
If the safe mode is enabled, Drainer modifies the replication updates in the following way:
```
